public/app/core/components/Select/FolderPicker.tsx

```diff
@@ -38,139 +42,83 @@ export interface Props {
   inputId?: string;
 }
+export type SelectedFolder = SelectableValue<number>;
+const VALUE_FOR_ADD = -10;
+
+export function FolderPicker(props: Props) {
+  const {
+    dashboardId,
+    allowEmpty,
+    onChange,
+    filter,
+    enableCreateNew,
+    inputId,
+    onClear,
+    enableReset,
+    initialFolderId,
+    initialTitle,
+    permissionLevel,
+    rootName,
+    showRoot,
+    skipInitialLoad,
+    accessControlMetadata,
+    customAdd,
+  } = props;
+  const isClearable = typeof onClear === 'function';
+  const [folder, setFolder] = useState<SelectedFolder | null>(null);
+  const [isCreatingNew, setIsCreatingNew] = useState(false);
+  const styles = useStyles2(getStyles);
+
+  const getOptions = useCallback(
+    async (query: string) => {
+      const searchHits = await searchFolders(query, permissionLevel, accessControlMetadata);
+      const options: Array<SelectableValue<number>> = mapSearchHitsToOptions(searchHits, filter);
+
+      const hasAccess =
+        contextSrv.hasAccess(AccessControlAction.DashboardsWrite, contextSrv.isEditor) ||
+        contextSrv.hasAccess(AccessControlAction.DashboardsCreate, contextSrv.isEditor);
+
+      if (hasAccess && rootName?.toLowerCase().startsWith(query.toLowerCase()) && showRoot) {
+        options.unshift({ label: rootName, value: 0 });
+      }
 
-interface State {
-  folder: SelectableValue<number> | null;
-}
-
-export class FolderPicker extends PureComponent<Props, State> {
-  debouncedSearch: any;
-
-  constructor(props: Props) {
-    super(props);
-
-    this.state = {
-      folder: null,
-    };
-
-    this.debouncedSearch = debounce(this.loadOptions, 300, {
-      leading: true,
-      trailing: true,
-    });
-  }
-
-  static defaultProps: Partial<Props> = {
-    rootName: 'General',
-    enableReset: false,
-    initialTitle: '',
-    enableCreateNew: false,
-    permissionLevel: PermissionLevelString.Edit,
-    allowEmpty: false,
-    showRoot: true,
-  };
-
-  componentDidMount = async () => {
-    if (this.props.skipInitialLoad) {
-      const folder = await getInitialValues({
-        getFolder: getFolderById,
-        folderId: this.props.initialFolderId,
-        folderName: this.props.initialTitle,
-      });
-      this.setState({ folder });
-      return;
-    }
-
-    await this.loadInitialValue();
-  };
-
-  // when debouncing, we must use the callback form of react-select's loadOptions so we don't
-  // drop results for user input. This must not return a promise/use await.
-  loadOptions = (query: string, callback: LoadOptionsCallback<number>): void => {
-    this.searchFolders(query).then(callback);
-  };
-
-  private searchFolders = async (query: string) => {
-    const {
-      rootName,
+      if (
+        enableReset &&
+        query === '' &&
+        initialTitle !== '' &&
+        !options.find((option) => option.label === initialTitle)
+      ) {
+        options.unshift({ label: initialTitle, value: initialFolderId });
```

Comment:
I'm not entirely sure about this piece of code, it seems to be prepending a blank option to the list (can also be seen in the video recording).

Can you elaborate on what it's checking for here that warrants a mutation to the options list?

Reply:
You were right @gillesdemey that causes a wrong empty row in the drop-down. I've just fixed:


https://user-images.githubusercontent.com/33540275/189921679-337d2203-9631-419a-af49-f5299557f76f.mp4